diff --git a/glslang/Include/Common.h b/glslang/Include/Common.h
index 9042a1aa..813d2296 100644
--- a/glslang/Include/Common.h
+++ b/glslang/Include/Common.h
@@ -140,15 +140,15 @@ namespace std {
     template<> struct hash<glslang::TString> {
         std::size_t operator()(const glslang::TString& s) const
         {
-            const unsigned _FNV_offset_basis = 2166136261U;
-            const unsigned _FNV_prime = 16777619U;
-            unsigned _Val = _FNV_offset_basis;
+            const unsigned foo_FNV_offset_basis = 2166136261U;
+            const unsigned foo_FNV_prime = 16777619U;
+            unsigned _Val = foo_FNV_offset_basis;
             size_t _Count = s.size();
             const char* _First = s.c_str();
             for (size_t _Next = 0; _Next < _Count; ++_Next)
             {
                 _Val ^= (unsigned)_First[_Next];
-                _Val *= _FNV_prime;
+                _Val *= foo_FNV_prime;
             }
 
             return _Val;
