diff --git a/BUILD.gn b/BUILD.gn
index bdcc4fc551..2545e7bd23 100644
--- a/BUILD.gn
+++ b/BUILD.gn
@@ -1061,6 +1061,7 @@ optional("gpu") {
   public_defines = []
   public_configs = []
   public_deps = []
+  public_include_dirs = []
 
   public = skia_gpu_public
   sources = skia_gpu_private
@@ -1085,7 +1086,9 @@ optional("gpu") {
         "src/gpu/ganesh/gl/egl/GrGLMakeEGLInterface.cpp",
         "src/gpu/ganesh/gl/egl/GrGLMakeNativeInterface_egl.cpp",
       ]
-      libs += [ "EGL" ]
+      # libs += [ "EGL" ]
+      public_include_dirs += [ "../angle/include" ]
+      public_defines += [ "KHRONOS_STATIC" ]
     } else if (skia_use_webgl) {
       sources += [ "src/gpu/ganesh/gl/webgl/GrGLMakeNativeInterface_webgl.cpp" ]
     } else if (is_linux && skia_use_x11) {
