diff --git a/src/deoptimizer/deoptimizer.h b/src/deoptimizer/deoptimizer.h
index e8b3a5bb8a7..c9920dcbcd6 100644
--- a/src/deoptimizer/deoptimizer.h
+++ b/src/deoptimizer/deoptimizer.h
@@ -114,6 +114,8 @@ class Deoptimizer : public Malloced {
   static bool IsDeoptimizationEntry(Isolate* isolate, Address addr,
                                     DeoptimizeKind* type_out);
 
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Winvalid-offsetof"
   // InstructionStream generation support.
   static int input_offset() { return offsetof(Deoptimizer, input_); }
   static int output_count_offset() {
@@ -124,6 +126,7 @@ class Deoptimizer : public Malloced {
   static int caller_frame_top_offset() {
     return offsetof(Deoptimizer, caller_frame_top_);
   }
+#pragma clang diagnostic pop
 
   Isolate* isolate() const { return isolate_; }
 
